File: problems/0006-zigzag-conversion/solution.py
"""
Problem Number: 6
Problem Name: Zigzag Conversion
Difficulty: Medium
Topics: String
Companies: PayPal, Amazon, Google
"""

class Solution:
    def convert(self, s: str, numRows: int) -> str:
        """
        Approach: Sort by Row (Simulation)
        Time Complexity: O(n)
        Space Complexity: O(n)
        """
        if numRows == 1 or numRows >= len(s):
            return s
        
        rows = [''] * numRows
        curRow, step = 0, 1
        
        for char in s:
            rows[curRow] += char
            if curRow == 0:
                step = 1
            elif curRow == numRows - 1:
                step = -1
            curRow += step
        
        return ''.join(rows)

# Test cases
if __name__ == "__main__":
    solution = Solution()
    print(solution.convert("PAYPALISHIRING", 3)) # "PAHNAPLSIIGYIR"
    print(solution.convert("PAYPALISHIRING", 4)) # "PINALSIGYAHRPI"
    print(solution.convert("A", 1))              # "A"

File: problems/0006-zigzag-conversion/solution.js
/**
 * @problemNumber: 6
 * @problemName: Zigzag Conversion
 * @difficulty: Medium
 * @topics: String
 * @companies: PayPal, Amazon, Google
 */

/**
 * Approach 1: Sort by Row (Simulation)
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 *
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
function convert(s, numRows) {
    if (numRows === 1 || s.length <= numRows) return s;

    const rows = new Array(numRows).fill('');
    let curRow = 0;
    let goingDown = false;

    for (let c of s) {
        rows[curRow] += c;
        if (curRow === 0 || curRow === numRows - 1) goingDown = !goingDown;
        curRow += goingDown ? 1 : -1;
    }

    return rows.join('');
}

// Test cases
console.log(convert("PAYPALISHIRING", 3)); // "PAHNAPLSIIGYIR"
console.log(convert("PAYPALISHIRING", 4)); // "PINALSIGYAHRPI"
console.log(convert("A", 1));              // "A"

module.exports = { convert };
